### DISCUSSION TITLE
Support for mermaid

### DISCUSSION THREAD

**Discussion by peters-r** [gethinode/hinode#944] [Help]
Hi,

first of all, it really is a great theme, thank you for providing this awesome theme!!

Sometimes I struggle a bit to understand the builtin magic.
I wanted to add mermaid support to my blog. What I did was copying the whole `baseof.html` file from the template to my hugo project and then I added the mermaid module as a script at the bottom of the page:

```
<script type="module">
  import mermaid from 'https://cdn.jsdelivr.net/npm/mermaid@10/dist/mermaid.esm.min.mjs';
</script>
```
And while this worked, it felt like the wrong way of doing it.
Another way is to put the script import into a partial, but also I have no Idea how to get my own partial into the footer without adding and overriding the original files from the theme.

Can you give me a hint what’s the right way of doing it? I read the docs, but I don’t understand this part and also I am not familiar with writing modules and somehow I hope I could avoid it.

Thank you and best regards,
Richard

**Comment by markdumay** [comment]
Glad you like the theme! I can imagine the theme's logic can be a bit challenging at first. But you're right, there's a more elegant way to integrate these kind of scripts. I'm not sure you're going to like it, because it does use Hugo's module system... ;-)

Hinode uses modules for the following reasons:
 - It provides a way to include them on only those pages that need it, reducing the overhead (and thus improving the site's performance)
 - It enables automated dependency upgrades, such as offered by GitHub's [Dependabot](https://docs.github.com/en/code-security/dependabot/working-with-dependabot)
 - It makes the codebase a bit more manageable

I prepared a guide [how to develop a Hugo module compatible with Hinode](https://gethinode.com/guides/modules/). It explains how to ingest JavaScript files and stylesheets into the build pipeline. The files are bundled into Hinode's core assets and minified to improve the site's performance. Hinode will also take care of (lazy) loading the files when applicable.

I'd be willing to include support for Mermaid diagrams as part of the Hinode ecosystem. Perhaps we could jointly develop such an open-source module?

**Reply by peters-r** [reply]
Hi Mark,

yes I would be happy to contribute. I could try to make a first draft. If I am stuck I will ask further questions here and than we can finalize together, what do you think?

There is a Hugo module: https://github.com/hugomods/mermaid but it doesn't support light and dark theme yet and also does not integrate with hinode well, since the partial does not get loaded automatically.

We could fork from there and extend it based on your mod-template example.

Should I fork it in my repository and we work there or do you want to fork it and then I make a PR to your repo? How would you like to do process wise, also in regard of how to pull it into the template later? Since you said you are willing to add support for mermaid I guess you would like to host the module, right?

Best regards,
Richard

**Reply by peters-r** [reply]
One aspect I dont understand yet is, it seems like you basically want to bundle all external javascript (as you also do with leaflet) into the build. But isn't it even better for performance and for caching if we also would support external sources (minified javascripts from CDN)?

So the part I am missing so far is how to pull in something like this (instead of providing the source of mermaid and bundling it):

```
<script type="module">
  import mermaid from 'https://cdn.jsdelivr.net/npm/mermaid@10/dist/mermaid.esm.min.mjs';
</script>
```

**Reply by markdumay** [reply]
Great! I'll create a new repository `gethinode/mod-mermaid` using `gethinode/mod-template` as template. This ensures the base skeleton for a Hinode module is in place. You can then submit PR's to a fork of this repo. The rest of your proposed approach / process makes sense to me. If you can make a first draft I will step in when needed.

I'd like to use dependabot and semantic release to automate our module as much as possible. Recently, I've created a (private) repo that uses [Husky](https://typicode.github.io/husky/), [commitlint](https://commitlint.js.org/), and [commitizen](https://github.com/commitizen/cz-cli). It ensures all commit messages are compatible with semantic release using native git hooks. Our `mod-template` would be nice open-source test case. If it all works, I plan to include this part in the `mod-template`.

**Reply by markdumay** [reply]
> One aspect I dont understand yet is, it seems like you basically want to bundle all external javascript (as you also do with leaflet) into the build. But isn't it even better for performance and for caching if we also would support external sources (minified javascripts from CDN)?

Hinode supports three types of integration points for JavaScripts. The critical files and core files are bundled and minified, whilst the optional files are included as individual files. 
 - **Critical** - scripts that have impact on the UI and should be loaded as soon as possible
 - **Core** - scripts that can be loaded lazily, but should be included on all pages (e.g. bootstrap, flex search)
 - **Optional** - scripts that can be loaded lazily, but should only be included on a page-by-page basis (e.g. leaflet)

The advantage of using a separate CDN for these kind of scripts when using static sites is limited:
 - It is an additional roundtrip to another end point
 - It is a potential security risk, as the source files are not within our control (increasing the attack vector)
 - It is difficult to automatically update and test version updates (as part of CI/CD)
 - Most hosting providers (e.g. Netlify, Cloudflare) are a CDN themselves, meaning our Hinode site including its script assets are deployed to the edge anyways.

**Reply by markdumay** [reply]
This thread provides some more background info: https://webmasters.stackexchange.com/questions/136816/for-a-website-behind-a-cdn-does-it-make-sense-to-combine-javascript-into-one-fi

Basically, modern browsers partition their caches, so there’s no performance gain across sites (like in the good old jquery days). Bundling isn’t recommended per se, as modern protocols support parallelization. You can use defer and async to lazily load the scripts.

**Reply by markdumay** [reply]
I primarily use bundling to control the loading order.

**Reply by peters-r** [reply]
Interesting, especially the partitioned caches, so this makes CDNs pointless for our use-case, you are right, so let's stick with your architecture. So mermaid would be an **optional** module, I think.

I am actually a noob in web development as I am mainly a Software Architect with embedded Linux background, so I might be a bit overwhelmed with the fancy toolchains you suggested, but I try contribute as much as I can and as I find time to :-)

**Reply by markdumay** [reply]
Sounds like a plan! Let’s take it one step at a time. The toolchains should be there to help us instead of hinder us... :-) I’ll set up a minimal repo first - probably tomorrow.

**Reply by markdumay** [reply]
I created a new repository: https://github.com/gethinode/mod-mermaid

Additionally, I followed the [module development steps](https://gethinode.com/docs/advanced-settings/module-development/#dependency-upgrades) to enable automation (the token configuration requires admin rights). We should now be able to submit PR's, test them, and automatically release a new version.

**Reply by markdumay** [reply]
As a note, the module version is now `v1.0.0`. Apparently, the first `feat` commit bumps the version to a major version, whilst it should still be an early `v0.x.x` development release. I'm still getting the hang of it. :-)

**Reply by peters-r** [reply]
Hi @markdumay, 

I adapted the exampleSite to use the hinode template. Unfortunately I get some build error, when I try to integrate mermaid.js with hinode. The error happens during bundling:

`executing "partials/footer/scripts.html" at <$js.Content>: error calling Content: EXECUTE-AS-TEMPLATE: failed to transform "/js/main.bundle.js" (text/javascript): failed to parse Resource "/js/main.bundle.js" as Template:: template: /js/main.bundle.js:33826: unexpected "\\" in command`

I get the same kind of message (but with different line), when I replace mermaid.js with katex.js, the one from your plugin. Are you familiar with this error message? It seems indeed to be related to specific lines in the javascript code, but I wonder why it also happens if I try with your katex.js file since the mod-katex plugin works on its own (I think).

I created a branch with my current work state:
https://github.com/peters-r/mod-mermaid/tree/develop/initial_mermaid_support

Thank you!

EDIT:

or if used as optional module:
`scripts.html:35:18": execute of template failed: template: partials/footer/scripts.html:35:18: executing "partials/footer/scripts.html" at <$js.Content>: error calling Content: EXECUTE-AS-TEMPLATE: failed to transform "/js/mermaid.bundle.js" (text/javascript): failed to parse Resource "/js/mermaid.bundle.js" as Template:: template: /js/mermaid.bundle.js:27116: unexpected "\\" in command`

The exact line in mermaid looks like this:
```defineMacro("\\notin", "\\html@mathml{\\mathrel{{\\in}\\mathllap{/\\mskip1mu}}}{\\mathrel{\\char`∉}}");```

And it's related to somthing in between the the part from `\\in` to `\\mathrel`.
If I remove the line than it hits somthing else later in the code with a similar problem and the affected line looks like:

`defineMacro("\\pmod", "\\pod{{\\rm mod}\\mkern6mu#1}");`

Even commenting out those lines doesn't help, so I guess it somehow statically analysis the javascript text and cannot deal with some tokens

**Reply by markdumay** [reply]
I encountered similar issues with the `katex` module. The scripts interfere with Hugo's processing pipeline (see [ExecuteAsTemplate](https://gohugo.io/hugo-pipes/resource-from-template/)).

I added the option [disableTemplate](https://gethinode.com/docs/configuration/modules/#configuring-modules) to exclude certain files / modules from the templates. Instead, Hinode will simply bundle the input as-is. Adding the following configuration to `params.toml` should fix the issue:

```
[modules]
    disableTemplate = ["katex", "mermaid"]
```

**Reply by peters-r** [reply]
Ah I see, yes thank you that solved this issue which was causing me a lot of trouble.

I would have one enhancement in mind to the core theme.

I feel like it doesn't scale well to put all disableTemplate modules into one array in the project configuration or the theme configuration.
Without this option, the module does not compile, but that's something the user shouldn't deal with and also the template shoudn't know about all possible modules in the future, probably.

I suggest to add all module configuration (and also their options for integration) rather into their own config section, so the modules can provide a default and the user can override of necessary, for example:

```toml
[modules]
    [modules.mermaid]
        integration = 'optional' // 'core' | 'critical' | 'optional'
        disableTemplate = true
        state = "defer"
        ....
```

As I unterstood the options are merged deeply (key-based), right?

What do you think?

**Reply by markdumay** [reply]
Good point, this approach would be much more scalable indeed. The options are merged deeply indeed, so you could still override them as needed. I'll submit a feature request so we can track this. I'm not sure if this would introduce a breaking change, or if we could use the current settings as fallback.

**Reply by markdumay** [reply]
I've created https://github.com/gethinode/hinode/issues/948.

**Reply by peters-r** [reply]
I would recommend keeping the fallback, but marking it as deprecated in the docs and to remove it in future major releases, if breaking changes will start to accumulate over time, so a new release could remove the fallbacks later

**Reply by peters-r** [reply]
@markdumay sorry I was a bit busy recently, but finally I made a PR: 

https://github.com/gethinode/mod-mermaid/pull/3

**Reply by markdumay** [reply]
Awesome work @peters-r! I've verified the code and merged it with the main module. Would you care to submit a PR against the Hinode core repo, so you would be mentioned as contributor?

Are there any missing features or items to work on? I checked the code examples and dark/light mode and didn't see any issues.

A final thought, would it make sense to strip the example site of `mod-mermaid` to the bare minimum? In theory, we now have a cyclic dependency between the `mod-mermaid` and `hinode` modules.

**Reply by peters-r** [reply]
Hi @markdumay, 

> Would you care to submit a PR against the Hinode core repo, so you would be mentioned as contributor?

Yes thank you that would be awesome!

> A final thought, would it make sense to strip the example site of mod-mermaid to the bare minimum?

Yes I will try, but I need to see how it looks like and if the plugin makes sense without the hinode core theme.

I will update within the upcoming week :-)

**Comment by peters-r** [comment]
BTW, sorry for adding an off-topic question...
It seems if I directly add markdown content to the _index.md (the front page) for an introduction text, then it does not seem to fill the whole width and is not centered (unlike the sections). I used sections instead, but they are not shown if there is no content besindes the index.md, so I currently don't have a good solution for a simple introduction text that aligns well with the layout.
